### ISSUE TITLE
Single YAML/properties key mapping from Secrets

### ISSUE THREAD

**Issue by Zetten** [spring-cloud/spring-cloud-kubernetes#841]
I'm working on an application where I need to map some list-of-object properties. In a traditional properties format:

```properties
example[0].foo=bar1
example[0].baz=qux1
example[1].foo=bar2
example[1].baz=qux2
```

Or yaml format;

```yaml
example:
- foo: bar1
  baz: qux1
- foo: bar2
  baz: qux2
```

Where previously I'd been using the simple key-value ConfigMap approach, this no longer works with indexed properties as the `[0]` syntax isn't valid for ConfigMap keys. Fortunately ConfigMap PropertySources can contain a single key describing properties- or yaml-formatted content, so I've reconfigured my deployment templates accordingly.

However, some of these properties are sensitive, and therefore should ideally be stored in Secrets. But the `SecretsPropertySource`/`SecretsPropertySourceLocator` don't have the same properties-/yaml-formatted property feature, and instead I end up with a single encoded property like:

```properties
application.properties=example[0].foo=bar1\nexample[1].baz=qux1\nexample[1].foo=bar2\nexample[1].baz=qux2
```

So of course my application doesn't get the actual values. This is true whether mounting as files in the Pod, or picked up via the API.

It seems effectively impossible to put such indexed property keys in Secrets and have them picked up by Spring Cloud Kubernetes.

I've now got crudely hacked versions of spring-cloud-kubernetes-commons and -fabric8-config which copy the logic from the `ConfigMapPropertySource*` classes over, and it all seems to work as expected.

But maybe I'm missing something - is there a reason that ConfigMaps and Secrets are treated so differently? Shouldn't the property parsing logic be consistent between the two?

**Comment by Zetten** [comment]
It may be useful to note that merging properties of objects-in-a-list doesn't actually work across multiple sources.

https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config

> When lists are configured in more than one place, overriding works by replacing the entire list.

In my particular case I can work around it by either a) binding all properties of a single object in a Secret, or b) creating a second 1:1-mapped list containing only the secret properties for each config object. In both cases I still would be using list-type mapping for Secrets.

**Comment by ryanjbaxter** [comment]
There is actually a PR to support this
https://github.com/spring-cloud/spring-cloud-kubernetes/pull/801/files

@wind57 I know you seem to have a strong opinion that this can be done in the application logic, but I don't see the harm in doing it in the property source.

**Comment by wind57** [comment]
@ryanjbaxter strong? oh, sorry if I made it sound like that, definitely not on that side. If so, totally my bad! As a matter of fact, I plan to take a few defects in config maps and secrets and fix them (like the PR above that you mention that lacks a bit traction), as soon as I am ready with my own little testing framework, so that I am sure I do not break anything and I am sure I exercise every single scenario. I _think_ this will happen in the upcoming weeks.

**Comment by Zetten** [comment]
That's great! Sorry, I did check for existing issues but missed that PR.

However, I think #801 misses the case where Secrets are mounted as volumes, which is handled directly in the [SecretsPropertySourceLocator](https://github.com/spring-cloud/spring-cloud-kubernetes/blob/main/spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/config/SecretsPropertySourceLocator.java#L91), and which collects each file as key:value :: filename:contents (i.e. the situation I described with Spring getting a property `application.properties` with value `<all the properties concatenated>`).
